cpp/CMakeLists.txt

```diff
@@ -429,7 +429,7 @@ add_library(cudf
             src/binaryop/jit/code/kernel.cpp
             src/binaryop/jit/code/operation.cpp
             src/binaryop/jit/code/traits.cpp
-            src/interop/from_arrow.cu
+            src/interop/from_arrow.cpp
```

Comment:
This change is to work around a CUDA 10.1 `nvcc` bug that only seems to happen on Ubuntu16.04. The Arrow change that exposed this bug is [here](https://github.com/apache/arrow/pull/7909/commits/239643e113346b36a3c73df983313fea3dcb1e04#diff-200337eb9b1aa7f7e176bdde8b3746d0R52-R56).

The fix in this commit splits `detail/concatenate.cuh` into two headers:
1. A `.hpp` file with the `detail::concatenate()` definitions (at `detail/concatenate.hpp`)
2. A `.cuh` file with the `detail::concatenate_masks()` definitions (at `detail/concatenate.cuh`)

The `detail/concatenate.cuh` header includes the new `detail/concatenate.hpp` so existing consumers of the `.cuh` header aren't affected.

Then `from_arrow.cu` is changed to include `detail/concatenate.hpp`. Then we rename `from_arrow.cu` -> `from_arrow.cpp`, so it's compiled by `gcc` instead of `nvcc`.

Here's the error we see in Ubuntu16.04/GCC7/CUDA 10.1:
```
include/arrow/array/array_nested.h:93:50: error: ‘arrow::BaseListArray’ is not a type
   friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
                                                  ^~~~~~~~~~~~~
include/arrow/array/array_nested.h: In instantiation of ‘class arrow::BaseListArray<arrow::ListType>’:
include/arrow/array/array_nested.h:103:63:   required from here
include/arrow/array/array_nested.h:93:13: error: template-id ‘SetListData<arrow::ListType>’ for ‘void arrow::internal::SetListData(int*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)’ does not match any template declaration
   friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
             ^~~~~~~~
include/arrow/array/array_nested.h:52:28: note: candidate is: template<class TYPE> void arrow::internal::SetListData(arrow::BaseListArray<TYPE>*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)
 template <typename TYPE>
                            ^          
include/arrow/array/array_nested.h: In instantiation of ‘class arrow::BaseListArray<arrow::LargeListType>’:
include/arrow/array/array_nested.h:147:68:   required from here
include/arrow/array/array_nested.h:93:13: error: template-id ‘SetListData<arrow::LargeListType>’ for ‘void arrow::internal::SetListData(int*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)’ does not match any template declaration
   friend void internal::SetListData<TYPE>(BaseListArray<TYPE>* self,
             ^~~~~~~~
include/arrow/array/array_nested.h:52:28: note: candidate is: template<class TYPE> void arrow::internal::SetListData(arrow::BaseListArray<TYPE>*, const std::shared_ptr<arrow::ArrayData>&, arrow::Type::type)
 template <typename TYPE>
                            ^          
```

CC: @rgsl888prabhu @harrism Is this an acceptable workaround for this issue?

Reply:
This is looks good, I don't see any issue.